sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -186,42 +217,48 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
-    var closedCoinsBoxesAmount : Long = 0L
-    var newCoinsBoxesAmount : Long = 0L
-
     if (!tx.isInstanceOf[MC2SCAggregatedTransaction]) {
+      validateProofsAndBoxValueAmount(tx)
+    }
 
-      for (u <- tx.unlockers().asScala) {
-        closedBox(u.closedBoxId()) match {
-          case Some(box) => {
-            val boxKey = u.boxKey()
-            if (!boxKey.isValid(box.proposition(), tx.messageToSign()))
-              throw new Exception("Box unlocking proof is invalid.")
-            if (box.isInstanceOf[CoinsBox[_ <: PublicKey25519Proposition]])
-              closedCoinsBoxesAmount += box.value()
-          }
-          case None => throw new Exception(s"Box ${u.closedBoxId()} is not found in state")
-        }
-      }
+    closedBoxesMerkleTree.validateTransaction(tx)
+
+    semanticValidity(tx).get
 
-      newCoinsBoxesAmount = tx.newBoxes().asScala
-        .filter(box => box.isInstanceOf[CoinsBox[_ <: PublicKey25519Proposition]] || box.isInstanceOf[WithdrawalRequestBox])
-        .map(_.value()).sum
+    if(!applicationState.validate(this, tx))
+      throw new Exception(s"ApplicationState transaction ${tx.id} validation failed.")
+  }
 
-      if (closedCoinsBoxesAmount != newCoinsBoxesAmount + tx.fee())
-        throw new Exception("Amounts sum of CoinsBoxes is incorrect. " +
-          s"ClosedBox amount - $closedCoinsBoxesAmount, NewBoxesAmount - $newCoinsBoxesAmount, Fee - ${tx.fee()}")
+  def validateProofsAndBoxValueAmount(tx: SidechainTypes#SCBT): Try[Unit] = Try {
+    var closedCoinsBoxesAmount : Long = 0L
+    var newCoinsBoxesAmount : Long = 0L
 
+    for (u <- tx.unlockers().asScala) {
+      closedBox(u.closedBoxId()) match {
+        case Some(box) => {
+          val boxKey = u.boxKey()
+          if (!boxKey.isValid(box.proposition(), tx.messageToSign()))
+            throw new Exception("Box unlocking proof is invalid.")
+          if (box.isInstanceOf[CoinsBox[_ <: PublicKey25519Proposition]])
+            closedCoinsBoxesAmount += box.value()
+        }
+        case None => throw new Exception(s"Box ${BytesUtils.toHexString(u.closedBoxId())} is not found in state")
+      }
     }
 
-    semanticValidity(tx).get
-    if(!applicationState.validate(this, tx))
-      throw new Exception(s"ApplicationState transaction ${tx.id} validation failed.")
+    newCoinsBoxesAmount = tx.newBoxes().asScala
+      .filter(box => box.isInstanceOf[CoinsBox[_ <: PublicKey25519Proposition]] || box.isInstanceOf[WithdrawalRequestBox])
+      .map(_.value()).sum
+
+    if (closedCoinsBoxesAmount != newCoinsBoxesAmount + tx.fee())
+      throw new Exception("Amounts sum of CoinsBoxes is incorrect. " +
+        s"ClosedBox amount - $closedCoinsBoxesAmount, NewBoxesAmount - $newCoinsBoxesAmount, Fee - ${tx.fee()}")
   }
 
   override def applyModifier(mod: SidechainBlock): Try[SidechainState] = {
     validate(mod).flatMap { _ =>
       changes(mod).flatMap(cs => {
+        closedBoxesMerkleTree.applyBlock(mod)
```

Comment:
Do we need to update `closedBoxesMerkleTree.applyBlock(mod)` directly here or inside `applyChanges(...)`. How will we react on unexpected failures?

Reply:
applicationState shall be checked first

Reply:
done